packages/codecs-core/src/__tests__/offset-codec-test.ts

```diff
@@ -0,0 +1,278 @@
+                mockCodec,
+                ({ preOffset }) => preOffset - 2,
+                ({ preOffset }) => preOffset,
+            ),
+        expectedPostOffset: 3,
+        expectedPreOffset: 1,
+        innerCodecSize: 2,
+        preOffset: 3,
+        size: 10,
+    },
+    {
+        _it: 'goes forwards and restores the original offset',
+        codec: mockCodec =>
+            offsetCodec(
+                mockCodec,
+                ({ preOffset }) => preOffset + 2,
+                ({ preOffset }) => preOffset,
+            ),
+        expectedPostOffset: 3,
+        expectedPreOffset: 5,
+        innerCodecSize: 2,
+        preOffset: 3,
+        size: 10,
+    },
+    {
+        _it: 'modulos the pre-offset',
+        codec: mockCodec => offsetCodec(mockCodec, ({ preOffset }) => preOffset + 100),
+        expectedPostOffset: 7,
+        expectedPreOffset: 3,
+        innerCodecSize: 4,
+        preOffset: 3,
+        size: 10,
+    },
+    {
+        _it: 'modulos the post-offset',
+        codec: mockCodec =>
+            offsetCodec(
+                mockCodec,
+                ({ preOffset }) => preOffset,
+                ({ postOffset }) => postOffset + 100,
+            ),
+        expectedPostOffset: 7,
+        expectedPreOffset: 3,
+        innerCodecSize: 4,
+        preOffset: 3,
+        size: 10,
+    },
+    {
+        _it: 'offsets using absolute positions',
+        codec: mockCodec =>
+            offsetCodec(
+                mockCodec,
+                () => 8,
+                () => 2,
+            ),
+        expectedPostOffset: 2,
+        expectedPreOffset: 8,
+        innerCodecSize: 4,
+        preOffset: 3,
+        size: 10,
+    },
+    {
+        _it: 'offsets using absolute positions modulo the byte length',
+        codec: mockCodec =>
+            offsetCodec(
+                mockCodec,
+                () => 18,
+                () => -6,
+            ),
+        expectedPostOffset: 4,
+        expectedPreOffset: 8,
+        innerCodecSize: 4,
+        preOffset: 3,
+        size: 10,
+    },
+    {
+        _it: 'does not modulo the post-offset if it has reached the end of the buffer',
+        codec: mockCodec =>
+            offsetCodec(
+                mockCodec,
+                () => 0,
+                ({ preOffset }) => preOffset,
+            ),
+        expectedPostOffset: 10,
+        expectedPreOffset: 0,
+        innerCodecSize: 4,
+        preOffset: 10,
+        size: 10,
+    },
+];
+
+describe('offsetCodec', () => {
+    it.each(testCases)(
+        '$_it',
+        ({ preOffset, size, innerCodecSize, codec: codecFactory, expectedPostOffset, expectedPreOffset }) => {
+            const mockCodec = getMockCodec({ size });
+            mockCodec.write.mockImplementation((_value, _bytes, offset) => offset + innerCodecSize);
+            mockCodec.read.mockImplementation((bytes, offset) => [bytes, offset + innerCodecSize]);
+
+            const codec = codecFactory(mockCodec);
```

Comment:
Inlining all of these test cases (as separate, hand-hewn artisanal `its()`) would reduce the debugging burden on the person who broke them. As it is, they're going to have to jump in and out of this code to figure out what the factory is, and what the expected values are.

Reply:
Thanks of the feedback. I've updated all the tests. Lmk what you think.